README.md
far_plugins
===========
This repository holds some plugins for [Far Manager][1].

* **LuaFAR for Editor** is an organizer of user's scripts and script packets.
* **LuaFAR History** manages histories of commands, folders, and viewed/edited files.

[1]: http://www.farmanager.com/


doc/changelog.txt
------------------------------------------------------------------
  Legend:   [+] added;   [-] deleted;   [*] changed;   [!] fixed;
------------------------------------------------------------------

2014-09-11, v3.6.0
  PLUGIN:
      [*] New syntax for Plugin.Call.
  Multi-Line Replace:
      [-] The utility moved to plugin LuaFAR Search.

2014-07-03, v3.5.3
  PLUGIN:
      [*] Rebuilt with up-to-date luaplug.c.
  Lua Expression:
      [!] The expression could not directly access global variables.

2014-03-27, v3.5.2
  PLUGIN:
      [*] Plugin links to lua51.dll rather than to lua5.1.dll.
  Multi-Line Replace:
      [!] Fix error message saying "nil".

2014-02-12, v3.5.1
  PLUGIN:
      [*] Plugin does not rely on LuaFAR anymore to include its path
          in package.path.

2013-10-07, v3.5.0
  PLUGIN:
      [*] Syntax of macro calls has changed again (run lfe: to see the syntax).

2013-07-26, v3.4.0
  PLUGIN:
      [*] Syntax of macro calls has changed:
          -- The 1-st argument must be string: one of the registered commands.
          -- There can be arbitrary number of additional arguments.
          -- There can be arbitrary number of returns.
      [-] Removed function far2.utils.CheckLuafarVersion.
  Block Sum:
      [*] Better alignment of inserted result.

2013-04-16, v3.3.1
  PLUGIN:
      [!] Fix jumps from error message window.
  Reformat Block:
      [*] "Process Lines": the line is deleted if the expression is false.

2013-04-02, v3.3.0
  PLUGIN:
      [*] Adapted for current LuaFAR API.
  Reformat Block:
      [*] "Prefix Lines" operation is replaced by much more powerful
          "Process Lines" one.

2013-03-22, v3.2.0
  PLUGIN:
      [*] Adapted for current LuaFAR API.

2012-12-16, v3.1.2
  Multi-Line Replace:
      [+] Added 'Count' operation.
  Script Parameters:
      [!] Fixed text centering on dialog separator.

2012-10-30, v3.1.1
      [*] Rebuild for current Far plugins API.

2012-09-29, v3.1.0
      [*] Rebuild for current Far plugins API.

2012-06-08, v3.0.8
      [+] Added utility: Multi-Line Replace.
      [-] Removed example: mergelines.lua.

2012-06-01, v3.0.7
      [+] Add 'Table View' utility (original author: Maxim Gonchar).
      [+] Add 'keynames.lua' module (original author: Aidar Rakhmatullin).
      [+] Examples added.

2012-04-09, v3.0.6
2012-03-26, v3.0.5
2012-03-24, v3.0.4
      [!] Maintenance versions (minor fixes).
      [+] Examples added.

2012-03-04, v3.0.3
  PLUGIN:
      [*] Function lf4ed.version moved to `utils' module and renamed to
          GetPluginVersion.

2012-01-04, v3.0.0
  PLUGIN:
      [*] Adaptation to Far 3.0 / LuaFAR 3.0 API.
      [*] Calling from macros via GUID instead of SysId.
      [-] "searchmenu" is not used by the plugin (but remains available
          to user scripts).
  Sort Lines:
      [*] Column pattern is Far regex (was: Lua regex).
      [*] Added flag SORT_STRINGSORT to case-insensitive sort.
      [+] "Case-sensitive" check boxes made 3-state.
  Block Sum:
      [+] If no text is selected, the current line is processed.
  Lua Script:
      [+] External script can be specified (via Script Parameters dialog).

2011-02-02, v2.8.0
  PLUGIN:
      [+] Got a SysId (0x10000) for calling from macros. The call syntax is
          the same as for command line calls (but the prefix is not needed).
          New switch -a for asynchronous calls.
      [+] Function MakeResident: parameter can be a table. That makes
          possible to require() files containing event handlers.
  Sort Lines:
      [+] a GUID added to the dialog.
  Reformat Block:
      [+] a GUID added to the dialog.

2010-12-24, v2.7.1
      [!] The build of 2.7.0 was broken due to an error in Makefile.

2010-12-23, v2.7.0
  PLUGIN:
      [+] 64-bit build added (works with 64-bit LuaFAR).
      [*] Plugin keeps its data in "%APPDATA%\LuaFAR for Editor" directory
          (was: in the plugin directory).

2010-11-26, v2.6.0
  PLUGIN:
      [*] Lua modules used by the plugin do not set global variables.
          Use the value returned by `require`.
  Sort Lines:
      [*] A single "undo" in Editor undoes the entire operation.
  Reformat Block:
      [*] A single "undo" in Editor undoes the entire operation.
  Block Sum:
      [*] The dialog moved from the configuration to the utility.
      [+] The result can be edited in the dialog.
      [*] Items immediately followed by [,;:] are considered valid.
  Lua Expression:
      [*] The dialog moved from the configuration to the utility.
      [+] The result can be edited in the dialog.
      [+] If there's no selection, the current line is processed.

2010-10-14, v2.5.0
  PLUGIN:
      [!] Error when calling "Block Sum" or "Lua Expression" settings dialogs.
      [!] Utilities could interfere with each other via parameters table.
  Lua Script:
      [+] Parameters can be passed to the script. Parameters dialog added.

2010-08-30, v2.4.0
  PLUGIN:
      [+] AddToMenu: parameter "where" can include letter "d" (for "dialog").
      [*] AddToMenu, AddCommand: unlimited number of additional arguments.
      [*] Installed scripts get their parameters in a table (was: 2 parameters).
          The table may also include fields "From" and "hDlg".
      [!] Fix "Reload user file" with default plugin settings.

2010-07-25, v2.3.0
  PLUGIN:
      [*] 'Reload User File' operation resets `package.loaded', in order to
          behave as close as possible to the initial loading.
      [+] new callback ("resident") function ExitScript.
      [+] function AddToMenu supports localization.

2010-06-25, v2.2.0
  PLUGIN:
      [*] LuaFAR 2.3 required.
      [+] Localization of the configuration dialog.
      [+] Utility for localization of script packets.
  Sort Lines:
      [!] Fixed shortcut keys in the dialogs.
      [!] Sorting made stable.
  Block Sum:
      [*] Made compatible with LuaFAR 2.3.

2010-04-24, v2.1.0
  PLUGIN:
      [!] Fixes to work with Unicode file names and paths (LuaFAR 2.2 required).

mreplace.lua
--local regex = require "rex_pcre"

local F=far.Flags
local AppName = "Multiline Replace"

local M = require "lf4ed_message"
local far2_dialog = require "far2.dialog"

local function ReplaceDialog (histData)
  local Dlg = far2_dialog.NewDialog()
  Dlg.dbox        = {"DI_DOUBLEBOX", 3, 1, 72,11, 0, 0, 0, 0, M.MMultilineReplace}
  Dlg.lab         = {"DI_TEXT",      5, 2,  0, 0, 0, 0, 0, 0, M.MDlgSearchPat}
  Dlg.sSearchPat  = {"DI_EDIT",      5, 3, 70, 4, 0, "SearchText", 0, {DIF_HISTORY=1,DIF_USELASTHISTORY=1}, ""}
  Dlg.lab         = {"DI_TEXT",      5, 4,  0, 0, 0, 0, 0, 0, M.MDlgReplacePat}
  Dlg.sReplacePat = {"DI_EDIT",      5, 5, 70, 0, 0, "ReplaceText", 0, {DIF_HISTORY=1,DIF_USELASTHISTORY=1}, ""}

  Dlg.bCaseSens   = {"DI_CHECKBOX",  5, 7,  0, 0, 0, 0, 0, 0, M.MDlgCaseSens}
  Dlg.bRegExpr    = {"DI_CHECKBOX", 26, 7,  0, 0, 0, 0, 0, 0, M.MDlgRegExpr}
  Dlg.bFileAsLine = {"DI_CHECKBOX", 48, 7,  0, 0, 0, 0, 0, 0, M.MDlgFileAsLine}
  Dlg.bWholeWords = {"DI_CHECKBOX",  5, 8,  0, 0, 0, 0, 0, 0, M.MDlgWholeWords}
  Dlg.bExtended   = {"DI_CHECKBOX", 26, 8,  0, 0, 0, 0, 0, 0, M.MDlgExtended}
  Dlg.bMultiLine  = {"DI_CHECKBOX", 48, 8,  0, 0, 0, 0, 0, 0, M.MDlgMultiline}

  Dlg.btnOk       = {"DI_BUTTON",    0,10,  0, 0, 0, 0, 0, {DIF_CENTERGROUP=1, DIF_DEFAULTBUTTON=1}, M.MOk}
  Dlg.btnCancel   = {"DI_BUTTON",    0,10,  0, 0, 0, 0, 0, F.DIF_CENTERGROUP, M.MCancel}

  local function CheckRegexChange (hDlg)
    local bRegex = Dlg.bRegExpr:GetCheck(hDlg)
    Dlg.bWholeWords :Enable(hDlg, not bRegex)
    Dlg.bExtended   :Enable(hDlg, bRegex)
    Dlg.bFileAsLine :Enable(hDlg, bRegex)
    Dlg.bMultiLine  :Enable(hDlg, bRegex)
  end

  local function DlgProc (hDlg, msg, param1, param2)
    if msg == F.DN_INITDIALOG then
      CheckRegexChange(hDlg)
    elseif msg == F.DN_BTNCLICK then
      if param1==Dlg.bRegExpr.id then CheckRegexChange(hDlg) end
    end
  end

  local id = win.Uuid("87ed8b17-e2b2-47d0-896d-e2956f396f1a")
  far2_dialog.LoadData(Dlg, histData)
  local ret = far.Dialog(id,-1,-1,76,13,"MReplace",Dlg,0,DlgProc)
  far2_dialog.SaveData(Dlg, histData)
  return ret==Dlg.btnOk.id
end

local function TransformSearchPattern (data)
  local pat = data.sSearchPat
  if not data.bRegExpr then
    pat = pat:gsub("[~!@#$%%^&*()%-+[%]{}\\|:;'\",<.>/?]", "\\%1")
    if data.bWholeWords then pat = "\\b"..pat.."\\b" end
  end
  return pat
end

local function TransformReplacePattern (data)
  if not data.bRegExpr then
    return data.sReplacePat
  end
  local subst = { ["\\"]="\\", ["%"]="%%", a="\a", e="\27", f="\f", n="\n", r="\r", t="\t" }
  local pat = regex.gsub(data.sReplacePat,
    [[ \\x([0-9a-f]{1,4}) | \\(.) | (\%) | \$([0-9]) ]],
    function(j,k,l,m)
      if j then
        j = unicode.utf8.char(tonumber(j,16))
        return j=="%" and "%%" or j
      end
      return k and (subst[k] or k) or l and "%%" or "%"..m
    end, nil, "ix")
  return pat
end

local function TransformFlags (data)
  local flags = data.bCaseSens and "" or "i"
  if data.bRegExpr then
    if data.bMultiLine  then flags=flags.."m" end
    if data.bFileAsLine then flags=flags.."s" end
    if data.bExtended   then flags=flags.."x" end
  end
  return flags
end

local function EditorAction (op, data)
  local editorInfo = editor.GetInfo()
  local bSelection = editorInfo.BlockType~=F.BTYPE_NONE
  local SearchPat = TransformSearchPattern(data)
  local ReplacePat = TransformReplacePattern(data)
  local Flags = TransformFlags(data)
  regex.gsub("foo", SearchPat, ReplacePat, nil, Flags) -- test patterns, provoke exception

  local t={}
  local lineno=bSelection and editorInfo.BlockStartLine or 0
  local eol
  editor.SetPosition(nil, lineno, 0)
  while lineno < editorInfo.TotalLines do
    local line=editor.GetString(nil, lineno)
    if bSelection and (line.SelStart<0 or line.SelEnd==0) then
      if eol~="" then t[#t+1]="" end
      break
    end
    t[#t+1], eol = line.StringText, line.StringEOL
    lineno = lineno+1
  end
  editor.SetPosition(nil, bSelection and editorInfo.BlockStartLine or 0, 0)

  local result, nFound, nReps = regex.gsub(table.concat(t,"\n"), SearchPat, ReplacePat, nil, Flags)
  if nFound == 0 then
    return nFound, nReps
  end

  editor.UndoRedo(nil, F.EUR_BEGIN)
  for i=bSelection and editorInfo.BlockStartLine or 0,lineno-1 do
    editor.DeleteString()
  end
  lineno = bSelection and editorInfo.BlockStartLine or 0
  for line, eol in result:gmatch("([^\r\n]*)(\r?\n?)") do
    if eol ~= "" then
      editor.InsertString()
      editor.SetString(nil, lineno, line)
      lineno = lineno+1
    else
      if line == "" then break end
      local L = editor.GetString()
      editor.SetString(nil, nil, line .. L.StringText)
    end
  end
  editor.UndoRedo(nil, F.EUR_END)

  return nFound, nReps
end

--------------------------------------------------------------------------------
------------------------- Test Suite BEGIN -------------------------------------
--------------------------------------------------------------------------------

local function OpenHelperEditor()
  local ret = editor.Editor ("__tmp__.tmp", nil, nil,nil,nil,nil,
              {EF_NONMODAL=1, EF_IMMEDIATERETURN=1, EF_CREATENEW=1}, 0, 0)
  assert (ret == F.EEC_MODIFIED, "could not open file")
end

local function CloseHelperEditor()
  editor.Quit()
  far.AdvControl("ACTL_COMMIT")
end

local function ProtectedError(msg, level)
  CloseHelperEditor()
  error(msg, level)
end

local function ProtectedAssert(condition, msg)
  if not condition then ProtectedError(msg or "assertion failed") end
end

local function GetEditorText()
  local t = {}
  editor.SetPosition(nil, 0, 0)
  for i=1, editor.GetInfo().TotalLines do
    t[i] = editor.GetString(nil, i-1, 2)
  end
  return table.concat(t, "\r")
end

local function SetEditorText(str)
  editor.SetPosition(nil,0,0)
  for i=1, editor.GetInfo().TotalLines do
    editor.DeleteString()
  end
  editor.InsertText(nil, str)
end

local function AssertEditorText(ref, msg)
  ProtectedAssert(GetEditorText() == ref, msg)
end

local function RunOneTest (lib, op, data, refFound, refReps)
  data.sRegexLib = lib
  editor.SetPosition(nil, data.CurLine or 0, data.CurPos or 0)
  local nFound, nReps = EditorAction(op, data)
  if nFound ~= refFound or nReps ~= refReps then
    ProtectedError(
      "nFound="        .. nFound..
      "; refFound="    .. refFound..
      "; nReps="       .. nReps..
      "; refReps="     .. refReps..
      "; sRegexLib="   .. tostring(data.sRegexLib)..
      "; bCaseSens="   .. tostring(data.bCaseSens)..
      "; bRegExpr="    .. tostring(data.bRegExpr)..
      "; bWholeWords=" .. tostring(data.bWholeWords)..
      "; bExtended="   .. tostring(data.bExtended)..
      "; bSearchBack=" .. tostring(data.bSearchBack)..
      "; sScope="      .. tostring(data.sScope)..
      "; sOrigin="     .. tostring(data.sOrigin)
    )
  end
end

local function test_Replace (lib)
  -- test empty replace
  local dt = { sSearchPat="l", sReplacePat="" }
  SetEditorText("line1\rline2\rline3\r")
  RunOneTest(lib, "test:replace", dt, 3, 3)
  AssertEditorText("ine1\rine2\rine3\r")

  -- test non-empty replace
  dt = { sSearchPat="l", sReplacePat="LL" }
  SetEditorText("line1\rline2\rline3\r")
  RunOneTest(lib, "test:replace", dt, 3, 3)
  AssertEditorText("LLine1\rLLine2\rLLine3\r")

  -- test submatches (captures)
  dt = { sSearchPat="(.)(.)(.)(.)(.)(.)(.)(.)(.)",
         sReplacePat="-$9-$8-$7-$6-$5-$4-$3-$2-$1-$0-",
         bRegExpr=true }
  local subj = "abcdefghi1234"
  SetEditorText(subj)
  RunOneTest(lib, "test:replace", dt, 1, 1)
  AssertEditorText("-i-h-g-f-e-d-c-b-a-abcdefghi-1234")

  -- test escaped dollar and backslash
  dt = { sSearchPat="abc", sReplacePat=[[$0\$0\t\\t]], bRegExpr=true }
  SetEditorText("abc")
  RunOneTest(lib, "test:replace", dt, 1, 1)
  AssertEditorText("abc$0\t\\t")

  -- test escape sequences in replace pattern
  local dt = { sSearchPat="b", sReplacePat=[[\a\e\f\n\r\t]], bRegExpr=true }
  for i=0,127 do dt.sReplacePat = dt.sReplacePat .. ("\\x%x"):format(i) end
  SetEditorText("abc")
  RunOneTest(lib, "test:replace", dt, 1, 1)
  local result = "a\7\27\12\13\13\9"
  for i=0,127 do result = result .. string.char(i) end
  result = result:gsub("\10", "\13", 1) .. "c"
  AssertEditorText(result)

  -- test replace in selection
  dt = { sSearchPat="in", sReplacePat="###", sScope="block" }
  SetEditorText("line1\rline2\rline3\rline4\r")
  editor.Select(nil, "BTYPE_STREAM",1,0,-1,2)
  RunOneTest(lib, "test:replace", dt, 2, 2)
  AssertEditorText("line1\rl###e2\rl###e3\rline4\r")

  -- test replace patterns containing \n or \r
  local dt = { sSearchPat=".", sReplacePat="a\rb", bRegExpr=true }
  dt.sOrigin = "scope"
  SetEditorText("L1\rL2\r")
  RunOneTest(lib, "test:replace", dt, 4, 4)
  AssertEditorText("a\rba\rb\ra\rba\rb\r")
  ------------------------------------------------------------------------------
end

local function Test()
  OpenHelperEditor()
  test_Replace("regex")
  CloseHelperEditor()
end

--------------------------------------------------------------------------------
------------------------- Test Suite END ---------------------------------------
--------------------------------------------------------------------------------

do
  local arg = ...
  if type(arg)=="table" and arg[1]=="dialog" then
    local histData = arg[2]
    if ReplaceDialog(histData) then
      local ok, ret = pcall(EditorAction, "replace", histData)
      editor.Redraw()
      if ok then
        local msg = (ret==0) and "No match found" or
          ("%d replacement%s made."):format(ret, ret==1 and "" or "s")
        far.Message(msg, AppName)
      else
        far.Message(ret, "Error", nil, "w")
      end
    end
  else
    Test()
    far.Message("All tests OK", AppName)
  end
end
